File: next/security/user and permission/backend-ms1/src/main/java/ma/zs/stocky/ws/converter/security/UserConverter.java
package  ma.zs.stocky.ws.converter.security;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import ma.zs.stocky.zynerator.util.ListUtil;

import ma.zs.stocky.ws.converter.security.RoleConverter;
import ma.zs.stocky.ws.converter.security.ModelPermissionUserConverter;
import ma.zs.stocky.ws.converter.security.ModelPermissionConverter;
import ma.zs.stocky.ws.converter.security.RoleUserConverter;
import ma.zs.stocky.ws.converter.security.ActionPermissionConverter;



import ma.zs.stocky.zynerator.util.StringUtil;
import ma.zs.stocky.zynerator.converter.AbstractConverter;
import ma.zs.stocky.zynerator.util.DateUtil;
import ma.zs.stocky.bean.core.security.User;
import ma.zs.stocky.ws.dto.security.UserDto;

@Component
public class UserConverter extends AbstractConverter<User, UserDto> {

    @Autowired
    private RoleConverter roleConverter ;
    @Autowired
    private ModelPermissionUserConverter modelPermissionUserConverter ;
    @Autowired
    private ModelPermissionConverter modelPermissionConverter ;
    @Autowired
    private RoleUserConverter roleUserConverter ;
    @Autowired
    private ActionPermissionConverter actionPermissionConverter ;
    private boolean roleUsers;
    private boolean modelPermissionUsers;

    public  UserConverter() {
        super(User.class, UserDto.class);
        init(true);
    }

    @Override
    public User toItem(UserDto dto) {
        if (dto == null) {
            return null;
        } else {
        User item = new User();
            if(StringUtil.isNotEmpty(dto.getId()))
                item.setId(dto.getId());
            if(dto.getCredentialsNonExpired() != null)
                item.setCredentialsNonExpired(dto.getCredentialsNonExpired());
            if(dto.getEnabled() != null)
                item.setEnabled(dto.getEnabled());
            if(StringUtil.isNotEmpty(dto.getEmail()))
                item.setEmail(dto.getEmail());
            if(dto.getAccountNonExpired() != null)
                item.setAccountNonExpired(dto.getAccountNonExpired());
            if(dto.getAccountNonLocked() != null)
                item.setAccountNonLocked(dto.getAccountNonLocked());
            if(StringUtil.isNotEmpty(dto.getUsername()))
                item.setUsername(dto.getUsername());
            if(StringUtil.isNotEmpty(dto.getPassword()))
                item.setPassword(dto.getPassword());
            if(dto.getPasswordChanged() != null)
                item.setPasswordChanged(dto.getPasswordChanged());
            if(StringUtil.isNotEmpty(dto.getLastName()))
                item.setLastName(dto.getLastName());
            if(StringUtil.isNotEmpty(dto.getFirstName()))
                item.setFirstName(dto.getFirstName());
            if(StringUtil.isNotEmpty(dto.getPhone()))
                item.setPhone(dto.getPhone());

            if(this.roleUsers && ListUtil.isNotEmpty(dto.getRoleUsers()))
                item.setRoleUsers(roleUserConverter.toItem(dto.getRoleUsers()));
            if(this.modelPermissionUsers && ListUtil.isNotEmpty(dto.getModelPermissionUsers()))
                item.setModelPermissionUsers(modelPermissionUserConverter.toItem(dto.getModelPermissionUsers()));


        return item;
        }
    }

    @Override
    public UserDto toDto(User item) {
        if (item == null) {
            return null;
        } else {
            UserDto dto = new UserDto();
            if(StringUtil.isNotEmpty(item.getId()))
                dto.setId(item.getId());
            if(StringUtil.isNotEmpty(item.getCredentialsNonExpired()))
                dto.setCredentialsNonExpired(item.getCredentialsNonExpired());
            if(StringUtil.isNotEmpty(item.getEnabled()))
                dto.setEnabled(item.getEnabled());
            if(StringUtil.isNotEmpty(item.getEmail()))
                dto.setEmail(item.getEmail());
            if(StringUtil.isNotEmpty(item.getAccountNonExpired()))
                dto.setAccountNonExpired(item.getAccountNonExpired());
            if(StringUtil.isNotEmpty(item.getAccountNonLocked()))
                dto.setAccountNonLocked(item.getAccountNonLocked());
            if(StringUtil.isNotEmpty(item.getUsername()))
                dto.setUsername(item.getUsername());
            if(StringUtil.isNotEmpty(item.getPassword()))
                dto.setPassword(item.getPassword());
            if(StringUtil.isNotEmpty(item.getPasswordChanged()))
                dto.setPasswordChanged(item.getPasswordChanged());
            if(StringUtil.isNotEmpty(item.getLastName()))
                dto.setLastName(item.getLastName());
            if(StringUtil.isNotEmpty(item.getFirstName()))
                dto.setFirstName(item.getFirstName());
            if(StringUtil.isNotEmpty(item.getPhone()))
                dto.setPhone(item.getPhone());
        if(this.roleUsers && ListUtil.isNotEmpty(item.getRoleUsers())){
            roleUserConverter.init(true);
            roleUserConverter.setUser(false);
            dto.setRoleUsers(roleUserConverter.toDto(item.getRoleUsers()));
            roleUserConverter.setUser(true);

        }
        if(this.modelPermissionUsers && ListUtil.isNotEmpty(item.getModelPermissionUsers())){
            modelPermissionUserConverter.init(true);
            modelPermissionUserConverter.setUser(false);
            dto.setModelPermissionUsers(modelPermissionUserConverter.toDto(item.getModelPermissionUsers()));
            modelPermissionUserConverter.setUser(true);

        }


        return dto;
        }
    }

    public void copy(UserDto dto, User t) {
    super.copy(dto, t);
    if (dto.getRoleUsers() != null)
        t.setRoleUsers(roleUserConverter.copy(dto.getRoleUsers()));
    if (dto.getModelPermissionUsers() != null)
        t.setModelPermissionUsers(modelPermissionUserConverter.copy(dto.getModelPermissionUsers()));
    }


    public void initList(boolean value) {
        this.roleUsers = value;
        this.modelPermissionUsers = value;
    }

    public void initObject(boolean value) {
    }


    public RoleConverter getRoleConverter(){
        return this.roleConverter;
    }
    public void setRoleConverter(RoleConverter roleConverter ){
        this.roleConverter = roleConverter;
    }
    public ModelPermissionUserConverter getModelPermissionUserConverter(){
        return this.modelPermissionUserConverter;
    }
    public void setModelPermissionUserConverter(ModelPermissionUserConverter modelPermissionUserConverter ){
        this.modelPermissionUserConverter = modelPermissionUserConverter;
    }
    public ModelPermissionConverter getModelPermissionConverter(){
        return this.modelPermissionConverter;
    }
    public void setModelPermissionConverter(ModelPermissionConverter modelPermissionConverter ){
        this.modelPermissionConverter = modelPermissionConverter;
    }
    public RoleUserConverter getRoleUserConverter(){
        return this.roleUserConverter;
    }
    public void setRoleUserConverter(RoleUserConverter roleUserConverter ){
        this.roleUserConverter = roleUserConverter;
    }
    public ActionPermissionConverter getActionPermissionConverter(){
        return this.actionPermissionConverter;
    }
    public void setActionPermissionConverter(ActionPermissionConverter actionPermissionConverter ){
        this.actionPermissionConverter = actionPermissionConverter;
    }
    public boolean  isRoleUsers(){
        return this.roleUsers ;
    }
    public void  setRoleUsers(boolean roleUsers ){
        this.roleUsers  = roleUsers ;
    }
    public boolean  isModelPermissionUsers(){
        return this.modelPermissionUsers ;
    }
    public void  setModelPermissionUsers(boolean modelPermissionUsers ){
        this.modelPermissionUsers  = modelPermissionUsers ;
    }
}

File: nest/security/backend-ms1/src/app/shared/converter/security/UserConverter.ts
import {forwardRef, Inject, Injectable} from "@nestjs/common";
import {AbstractConverter} from "src/app/zynerator/converter/AbstractConverter";

import {User} from "src/app/shared/bean/core/security/User";
import {UserDto} from "src/app/shared/dto/security/UserDto";

import {RoleConverter} from "src/app/shared/converter/security/RoleConverter";
import {ModelPermissionUserConverter} from "src/app/shared/converter/security/ModelPermissionUserConverter";
import {ModelPermissionConverter} from "src/app/shared/converter/security/ModelPermissionConverter";
import {RoleUserConverter} from "src/app/shared/converter/security/RoleUserConverter";
import {ActionPermissionConverter} from "src/app/shared/converter/security/ActionPermissionConverter";

@Injectable()
export class UserConverter extends AbstractConverter<User, UserDto> {

    roleUsers: boolean;
    modelPermissionUsers: boolean;

    constructor(
        @Inject(forwardRef(() => RoleConverter)) readonly roleConverter: RoleConverter,
        @Inject(forwardRef(() => ModelPermissionUserConverter)) readonly modelPermissionUserConverter: ModelPermissionUserConverter,
        @Inject(forwardRef(() => ModelPermissionConverter)) readonly modelPermissionConverter: ModelPermissionConverter,
        @Inject(forwardRef(() => RoleUserConverter)) readonly roleUserConverter: RoleUserConverter,
        @Inject(forwardRef(() => ActionPermissionConverter)) readonly actionPermissionConverter: ActionPermissionConverter,
        ) {
            super();
            this.init(true);
    }

    toItem(dto: UserDto): User {
        if (!dto) {
            return null;
        }
        const item =new User();
        if (dto.id) {
            item.id = dto.id;
        }
        if (dto.credentialsNonExpired) {
            item.credentialsNonExpired = dto.credentialsNonExpired;
        }
        if (dto.enabled) {
            item.enabled = dto.enabled;
        }
        if (dto.email) {
            item.email = dto.email;
        }
        if (dto.accountNonExpired) {
            item.accountNonExpired = dto.accountNonExpired;
        }
        if (dto.accountNonLocked) {
            item.accountNonLocked = dto.accountNonLocked;
        }
        if (dto.username) {
            item.username = dto.username;
        }
        if (dto.password) {
            item.password = dto.password;
        }
        if (dto.passwordChanged) {
            item.passwordChanged = dto.passwordChanged;
        }
        if (dto.lastName) {
            item.lastName = dto.lastName;
        }
        if (dto.firstName) {
            item.firstName = dto.firstName;
        }
        if (dto.phone) {
            item.phone = dto.phone;
        }



		if(this.roleUsers && dto.roleUsers)
			item.roleUsers = this.roleUserConverter.toItems(dto.roleUsers);
		if(this.modelPermissionUsers && dto.modelPermissionUsers)
			item.modelPermissionUsers = this.modelPermissionUserConverter.toItems(dto.modelPermissionUsers);

        return item;
    }

    toDto(item: User): UserDto {
        if (!item) {
            return null;
        }
        const dto = new UserDto();

        if (item.id) {
            dto.id = item.id;
        }
        if (item.credentialsNonExpired) {
            dto.credentialsNonExpired = item.credentialsNonExpired;
        }
        if (item.enabled) {
            dto.enabled = item.enabled;
        }
        if (item.email) {
            dto.email = item.email;
        }
        if (item.accountNonExpired) {
            dto.accountNonExpired = item.accountNonExpired;
        }
        if (item.accountNonLocked) {
            dto.accountNonLocked = item.accountNonLocked;
        }
        if (item.username) {
            dto.username = item.username;
        }
        if (item.password) {
            dto.password = item.password;
        }
        if (item.passwordChanged) {
            dto.passwordChanged = item.passwordChanged;
        }
        if (item.lastName) {
            dto.lastName = item.lastName;
        }
        if (item.firstName) {
            dto.firstName = item.firstName;
        }
        if (item.phone) {
            dto.phone = item.phone;
        }

        if(this.roleUsers && item.roleUsers?.length > 0){
            this.roleUserConverter.init(true);
            this.roleUserConverter.user = false;
            dto.roleUsers = this.roleUserConverter.toDtos(item.roleUsers);
            this.roleUserConverter.user = true;
        }
        if(this.modelPermissionUsers && item.modelPermissionUsers?.length > 0){
            this.modelPermissionUserConverter.init(true);
            this.modelPermissionUserConverter.user = false;
            dto.modelPermissionUsers = this.modelPermissionUserConverter.toDtos(item.modelPermissionUsers);
            this.modelPermissionUserConverter.user = true;
        }

        return dto;
    }


    public initList(value: boolean): void {
            this.roleUsers = value;
            this.modelPermissionUsers = value;
    }

    public initObject(value: boolean): void {
    }
}
